### PR TITLE
Filter list features

### PR DESCRIPTION
**Pull request by saulyz**
There are several minor problems to be solved soon:
- Filter form dialog clears `breed` if this filter is still set
- some optimization is needed (todo comments)
- console logs

### CODE REVIEW

File: src/components/animal/AnimalsListContainer.tsx [hunk 1]
```diff
@@ -64,54 +50,29 @@ export default function AnimalsListContainer({ viewType, setAnimalsCount }: Anim
...
     }
 
     function handlePageSizeChange(size) {
-        setCurrentPage(0);
-        setPageSize(size);
-        loadFirstPage(size);
+        dispatch(setAnimalsPageSize(size));
     }
 
     function handlePageChange(newPage) {
```
Comment (rikius) [inline src/components/animal/AnimalsListContainer.tsx:69]: for me it looks like this function is redundant in this component.

File: src/components/animal/create-update/AnimalForm.tsx [hunk 2]
```diff
@@ -33,7 +35,16 @@ export default function AnimalForm({ animal }: AnimalFormProps) {
         if (animal) {
             formData.id = animal.id;
         }
-        dispatch(createOrUpdateAnimal(normalizedFormData));
+        dispatch(createOrUpdateAnimal(normalizedFormData, submitCallback));
+    };
+
+    const submitCallback = (error: any) => {
```
Comment (rikius) [inline src/components/animal/create-update/AnimalForm.tsx:41]: form should not decide where to redirect user, upper component have to do it.

File: src/components/animal/filters/AnimalFilterChips.tsx [hunk 3]
```diff
@@ -1,40 +1,128 @@
...
+    const dispatch = useDispatch();
+    const [filters, setFilters] = useState<Filter[]>([]);
+    const { queryVars, queryVarsFilterObjs } = useSelector((state: RootStateOrAny) => state.animalsAll);
+
+    const hasFilters = Object.keys(queryVars).filter(key => FILTER_PROPS.includes(key)).length > 0;
+
+    useEffect(() => {
+        setFilters(makeFilterValues(queryVars, queryVarsFilterObjs));
+        // eslint-disable-next-line react-hooks/exhaustive-deps
+    }, [queryVars]);
```
Comment (rikius) [inline src/components/animal/filters/AnimalFilterChips.tsx:20]: it should be both object: queryVarsFilterObjs and queryVars or just one in:
```        
setFilters(makeFilterValues(queryVars, queryVarsFilterObjs));
```

File: src/components/animal/filters/AnimalFilterChips.tsx [hunk 4]
```diff
@@ -1,40 +1,128 @@
...
-        const newFilters = [...filters];
-        newFilters.forEach(filter => {
-            filter.value = undefined;
-            filter.displayValue = undefined;
+        dispatch(setAnimalsFiltersWithFilterObjs({}));
+    };
+
+    const onDelete = (filter: Filter) => {
+        // todo - remove
+        console.log('onDelete', filter, JSON.stringify(queryVarsFilterObjs));
```
Comment (rikius) [inline src/components/animal/filters/AnimalFilterChips.tsx:28]: production code should not have console log outputs.

File: src/components/animal/filters/AnimalFilterChips.tsx [hunk 5]
```diff
@@ -1,40 +1,128 @@
...
+
+        const newFilterObjs = {};
+        FILTER_PROPS.map(type => {
+            const filterPropList = queryVarsFilterObjs[type];
+            const filterPropValuesAfterDelete = !filterPropList
+                ? null
+                : filterPropList.filter(prop => {
+                      // todo - remove
+                      console.log('in', type, 'filter with', prop, filter);
```
Comment (rikius) [inline src/components/animal/filters/AnimalFilterChips.tsx:37]: production code should not have console log outputs.

File: src/components/animal/filters/AnimalFilterChips.tsx [hunk 6]
```diff
@@ -1,40 +1,128 @@
...
         <Grid container spacing={1}>
-            {filters
-                .filter(filter => filter.value)
-                .map(filter => (
-                    <Grid item key={filter.name}>
-                        <Chip label={filter.displayValue} onDelete={() => onDelete(filter)} />
-                    </Grid>
-                ))}
-            <Button onClick={handleClearFilters}>Clear</Button>
+            {filters.map(filter => (
```
Comment (rikius) [inline src/components/animal/filters/AnimalFilterChips.tsx:51]: why removed filtering for empty values?

File: src/components/animal/filters/AnimalFilterChips.tsx [hunk 7]
```diff
@@ -1,40 +1,128 @@
...
-    onDelete: (filter: Filter) => void;
-    onClearFilters: (filters: Filter[]) => void;
+const makeFilterValues = (queryVars, queryVarsFilterObjs) => {
+    // todo - remove
+    console.log('makeFilterValues', JSON.stringify(queryVars), JSON.stringify(queryVarsFilterObjs));
+    const filters: Filter[] = [];
+    const { species: speciesObjList, breed: breedObjList, gender: genderObjList } = queryVarsFilterObjs;
+
+    // todo - optimize
+    if (queryVars.species) {
```
Comment (rikius) [inline src/components/animal/filters/AnimalFilterChips.tsx:68]: yes need to optimize it multiple similar code places.

File: src/components/animal/filters/AnimalFilterChips.tsx [hunk 8]
```diff
@@ -1,40 +1,128 @@
...
+const getFilterPropValue = (id: number, dataset: Species[] | Breed[] | Gender[]): string | boolean => {
+    if (!dataset || dataset.length === 0) {
+        return false;
+    }
+    const genericDataset = dataset as GenericDataset[];
+    const found = genericDataset.find(item => item.id === id);
+    if (!found) {
+        return false;
+    }
+    return found?.value ?? 'unknown';
```
Comment (rikius) [inline src/components/animal/filters/AnimalFilterChips.tsx:116]: unknown item will be displayed in the list. better is not to make a filter according this value at all.

File: src/components/animal/filters/AnimalFiltersDialog.tsx [hunk 9]
```diff
@@ -0,0 +1,128 @@
...
+    };
+
+    const onResetFilters = () => {
+        reset({});
+        dispatch(setAnimalsFiltersWithFilterObjs({}));
+    };
+
+    const onSubmit = formData => {
+        handleDialogClose();
+        console.log('onSubmit with raw formData', formData);
```
Comment (rikius) [inline src/components/animal/filters/AnimalFiltersDialog.tsx:55]: remove console log.

File: src/components/animal/filters/AnimalFiltersDialog.tsx [hunk 10]
```diff
@@ -0,0 +1,128 @@
...
+        const preparedFilterObjs = makeMultipleFromSingleFilterProps(formData);
+        console.log('onSubmit with', preparedFilterObjs);
```
Comment (rikius) [inline src/components/animal/filters/AnimalFiltersDialog.tsx:57]: remove console.log

File: src/components/animal/filters/AnimalFiltersDialog.tsx [hunk 11]
```diff
@@ -0,0 +1,128 @@
...
+
+    const onResetFilters = () => {
+        reset({});
+        dispatch(setAnimalsFiltersWithFilterObjs({}));
+    };
+
+    const onSubmit = formData => {
+        handleDialogClose();
+        console.log('onSubmit with raw formData', formData);
+        const preparedFilterObjs = makeMultipleFromSingleFilterProps(formData);
```
Comment (rikius) [inline src/components/animal/filters/AnimalFiltersDialog.tsx:56]: use deconstruction of object to needed variables.

File: src/components/animal/AnimalsListContainer.tsx (hunk 1 above)
Comment (saulyz) [reply src/components/animal/AnimalsListContainer.tsx:69]: I had this feeling.  Management of the current page is now moved to Store internals.

File: src/components/animal/filters/AnimalFilterChips.tsx (hunk 3 above)
Comment (saulyz) [reply src/components/animal/filters/AnimalFilterChips.tsx:20]: Here we have a problem that queryVars and queryVarsFilterObjs changes in different time. 
The main is queryVars and queryVarsFilterObjs is only for UI (change of queryVarsFilterObjs does not make an impact)

**Review by rikius — APPROVED** [review]

**Pull request merged**